contracts/ERC721FreezeVoting.sol

```diff
@@ -0,0 +1,105 @@
+//SPDX-License-Identifier: MIT
+pragma solidity =0.8.19;
+
+import { IERC721VotingStrategy } from "./azorius/interfaces/IERC721VotingStrategy.sol";
+import { BaseFreezeVoting, IBaseFreezeVoting } from "./BaseFreezeVoting.sol";
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+/**
+ * A [BaseFreezeVoting](./BaseFreezeVoting.md) implementation which handles 
+ * freezes on ERC721 based token voting DAOs.
+ */
+contract ERC721FreezeVoting is BaseFreezeVoting {
+
+    /** A reference to the voting strategy of the parent DAO. */
+    IERC721VotingStrategy public strategy;
+
+    /**
+     * Mapping of block the freeze vote was started on, to the token address, to token id,
+     * to whether that token has been used to vote already.
+     */
+    mapping(uint256 => mapping(address => mapping(uint256 => bool))) public idHasFreezeVoted;
+
+    event ERC721FreezeVotingSetUp(address indexed owner, address indexed strategy);
+
+    error NoVotes();
+    error NotSupported();
+    error UnequalArrays();
+
+    /**
+     * Initialize function, will be triggered when a new instance is deployed.
+     *
+     * @param initializeParams encoded initialization parameters.
+     */
+    function setUp(bytes memory initializeParams) public override initializer {
+        (
+            address _owner,
+            uint256 _freezeVotesThreshold,
+            uint32 _freezeProposalPeriod,
+            uint32 _freezePeriod,
+            address _strategy
+        ) = abi.decode(
+                initializeParams,
+                (address, uint256, uint32, uint32, address)
+            );
+
+        __Ownable_init();
+        _transferOwnership(_owner);
+        _updateFreezeVotesThreshold(_freezeVotesThreshold);
+        _updateFreezeProposalPeriod(_freezeProposalPeriod);
+        _updateFreezePeriod(_freezePeriod);
+        freezePeriod = _freezePeriod;
+        strategy = IERC721VotingStrategy(_strategy);
+
+        emit ERC721FreezeVotingSetUp(_owner, _strategy);
+    }
+
+    function castFreezeVote() external override pure { revert NotSupported(); }
+
+    function castFreezeVote(address[] memory _tokenAddresses, uint256[] memory _tokenIds) external {
+        if (_tokenAddresses.length != _tokenIds.length) revert UnequalArrays();
+
+        uint256 userVotes = _getVotesAndUpdateHasVoted(_tokenAddresses, _tokenIds, msg.sender);
```

Comment:
So this line is getting called before `freezeProposalCreatedBlock` is getting updated, so it could be counting user votes from the wrong block right?

Reply:
dammit, yeah good catch

I want to have the update be in the same look so I don't need to do it twice, but that might be tricky